tests/interfaces/test_autograd.py

```diff
@@ -515,10 +541,18 @@ def circuit(weights, data1, data2):
         assert len(res) == 1
         assert res[0].shape == weights.shape
 
-        # check that the parameter shift was only performed for the
-        # differentiable elements of `weights`, not the data input
+        # check that the first arg was marked as non-differentiable
         assert circuit.get_trainable_args() == {0}
-        assert _inv_dict(circuit.par_to_grad_method)["0"] == {18, 19, 20, 21, 22, 23}
+
+        # Check that the parameter shift was not performed for the
+        # non-differentiable elements of `data1` and `data2`.
+        # First, extract the variable indices that the jacobian method
+        # 'skipped' (those with grad_method="0"):
```

Comment:
Was it always the case that "0" was a valid `grad_method`? :thinking:

Reply:
Kind of. It's assigned by the QNode to parameters, rather than an operation grad method.

For example, prior to this refactor, if you had a QNode with a parameter that is never used:

```python
@qml.qnode(dev)
def circuit(a, b):
    qml.RX(a, wires=0)
    return qml.expval(qml.PauliZ(0))
```

then to avoid a pointless parameter-shift occurring for parameter `b`, the QNode would assign it a grad method of `"0"` and automatically insert 0's into the Jacobian rather than call the parameter shift method.

Reply:
Here, I'm just taking advantage of this pre-existing logic to also assign `"0"` to non-differentiable QNode arguments, even if they are used in operations.